x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -0,0 +1,273 @@
+            try (IndexInput input = in.openInput(fileName, ioContext)) {
+                if (start > 0) {
+                    input.seek(start);
+                }
+                int bytesCopied = 0;
+                long remaining = end - start;
+                while (remaining > 0) {
+                    final int size = (remaining < copyBuffer.length) ? Math.toIntExact(remaining) : copyBuffer.length;
+                    input.readBytes(copyBuffer, 0, size);
+                    fc.write(ByteBuffer.wrap(copyBuffer, 0, size), start + bytesCopied);
+                    bytesCopied += size;
+                    remaining -= size;
+                }
+            }
+        }
+
+        @Override
+        protected void seekInternal(long pos) throws IOException {
+            if (pos > length()) {
+                throw new EOFException("Reading past end of file [position=" + pos + ", length=" + length() + "] for " + toString());
+            } else if (pos < 0L) {
+                throw new IOException("Seeking to negative position [" + pos + "] for " + toString());
+            }
+        }
+
+        @Override
+        public CacheBufferedIndexInput clone() {
+            final CacheBufferedIndexInput clone = (CacheBufferedIndexInput) super.clone();
+            clone.cacheFile = new AtomicReference<>();
+            clone.closed = new AtomicBoolean(false);
+            clone.parent = (this.parent != null ? this.parent : this);
+            assert clone.parent.parent == null : "parent must not be a clone";
+            return clone;
+        }
+
+        @Override
+        public IndexInput slice(String sliceDescription, long offset, long length) throws IOException {
+            if (offset < 0 || length < 0 || offset + length > this.length()) {
+                throw new IllegalArgumentException("slice() " + sliceDescription + " out of bounds: offset=" + offset
+                    + ",length=" + length + ",fileLength=" + this.length() + ": " + this);
+            }
+            final CacheBufferedIndexInput slice = new CacheBufferedIndexInput(fileName, fileLength, ioContext,
+                getFullSliceDescription(sliceDescription), this.offset + offset, length);
+            slice.parent = (this.parent != null ? this.parent : this);
+            assert slice.parent.parent == null : "parent must not be a clone";
+            return slice;
+        }
+
+        @Override
+        public String toString() {
+            return "CacheBufferedIndexInput{" +
+                "fileName='" + fileName + '\'' +
+                ", fileLength=" + fileLength +
+                ", offset=" + offset +
+                ", end=" + end +
+                ", length=" + length() +
+                ", clone=" + (parent != null) +
+                ", position=" + getFilePointer() +
+                '}';
+        }
+
+        private int copySource(long start, long end, byte[] buffer, int offset) throws IOException {
```

Comment:
Suggest renaming to something like `readDirectly`

Reply:
This is a good suggestion, thanks. I pushed https://github.com/elastic/elasticsearch/pull/50693/commits/391633f5150655187589c25a3b4a1c2f706587cf